### DISCUSSION TITLE
84. MVCC에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#86] [General]
MVCC에 대해 설명해 주세요.

**Comment by linirini** [comment]
MVCC(Multi-Version Concurrency Control)는 트랜잭션 간의 동시성을 보장하면서 락을 최소화하고 성능을 향상시키기 위해 사용하는 기법이다. MVCC는 데이터를 수정할 때 여러 버전을 생성하여 읽기 작업과 쓰기 작업 간 충돌을 줄이고, 트랜잭션의 일관성과 고립성을 보장한다.
### MVCC의 핵심 아이디어
- 데이터의 다중 버전 관리
  - 데이터를 변경할 때 기존 데이터를 바로 덮어쓰지 않고, 새로운 버전을 생성한다.
  - 각 트랜잭션은 트랜잭션 시작 시점의 데이터 스냅샷을 기반으로 작업을 수행한다.
- 읽기와 쓰기의 분리
  - 읽기 작업은 데이터의 특정 버전을 읽고, 쓰기 작업은 새 버전을 생성한다.
  - 읽기 작업은 다른 트랜잭션의 쓰기 작업에 방해받지 않는다.
- 트랜잭션의 타임스탬프
  - 각 트랜잭션의 데이터 버전은 타임 스탬프나 고유 식별자로 관리
### MVCC의 작동 원리
1. 트랜잭션의 읽기 작업
  - 스냅샷 읽기
    - 트랜잭션은 자신이 시작된 시점에서 볼 수 있는 데이터의 스냅샷을 사용
    - 다른 트랜잭션이 데이터를 변경하더라도, 트랜잭션 시작 시점의 데이터 버전을 참조하므로 일관된 읽기 제공
2. 트랜잭션의 쓰기 작업
  - 새 버전 생성
    - 데이터를 수정할 때 기존 데이터를 직접 변경하지 않고 새로운 데이터 버전 생성
    - 수정 작업 완료 후 다른 트랜잭션은 이 버전을 참조
3. 가시성 판단
  - 각 데이터 버전은 다음 정보를 포함
    - 생성 트랜잭션 ID: 데이터를 생성한 트랜잭션
    - 삭제 트랜잭션 ID: 해당 시점의 데이터가 무효화된 경우에
  - 읽는 트랜잭션은 자신과 관련 없는 트랜잭션이 생성한 데이터 버전을 무시한다.
### MVCC 장점
- 동시성 향상
- 일관성 보장
- 읽기 성능 최적화
### MVCC 단점
- 공간 효율성 저하
- 복잡성 증가
- 쓰기 성능 저하

****MySQL(InnoDB)**
- MVCC를 기반으로 트랜잭션의 격리 수준을 제공
- Undo Log를 활용하여 이전 버전을 관리하고, 읽기 작업은 Undo Log에서 필요한 데이터 버전을 참조
